keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,170 @@
+---
+kep-number: 36
+title: Node Failure Recovery with ForceEviction Taint
+authors:
+  - "@jingxu97"
+  - "@yastij"
+owning-sig: sig-node
+participating-sigs:
+  - sig-storage
+  - sig-node
+  - sig-architecture
+reviewers:
+  - "@yujuhong"
+  - "@yguo00905"
+  - "@liggitt"
+approvers:
+  - "@liggitt"
+  - "@yujuhong"
+editor: TBD
+creation-date: 2019-1-18
+last-updated: 2019-1-18
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+Node Failure Recovery with ForceEviction Taint
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+To address the issue of failing to recove Pod from node failure, we are proposing a new
+taint-base eviction effect, ForceEviction. With this new type of eviction, Pods can be
+deleted successfully without waiting for Kubelet's acknowldegement and volumes can be
+detached faster. The main challenge of this proposal is to make sure no race condition
+between Pod eviction and Kubelet restarts Pods when it comes back.
+
+
+## Motivation
+
+Kubernetes relies on a node agent (aka kubelet) to report the health of a node. 
+There are a variety of reasons that kubelet fails to update its node status,
+such as kernel crash, panic, or update which causes system reboot,
+system hangs due to heavy load, network partition, and machine is powered off
+due to failures. When this happens, Kubernetes currently uses a taint-based eviction
+trying to recovery from the faiued node. The recovery process will taint the node
+status as “NotReady” and will trigger eviction of pods on the failed node if the node
+is in "NotReady" state for a while (set to 5 minutes by default) so that new pods
+could be started on other healthy nodes. 
+
+However, users often reported issues during the failure handling including old pods
+cannot be deleted correctly, new pods cannot be started correctly due to volume failing
+to detach from old node and attach to new node issues, or it takes a long time for new
+pods to start. The main reason for this problem is due to the fact that Pod deletion requires
+kubelet's acknowlegement. When a Pod is eviceted (deleted) from a node, the deletion timestamp
+is first set on the Pod. However the Pod will not be actually deleted from API server until kubelet 
+finishes cleaning up all the resources of this Pod and kills the running containers. If the node
+is in a failed state and kubelet cannot respond, Pod will be stuck in a teminating state forever.
```

Comment:
Just for accuracy - Pods don't get stuck "forever". They do eventually get deleted, it may take more than couple of hours though.

Reply:
I am not sure about this. Why it gets deleted after a couple of hours if kubelet is not back. I think only Kubelet can eventfully force delete the pod.